src/model/implementation/i18n.ts

```diff
@@ -0,0 +1,285 @@
+            switch (providerName) {
+                case I18N_GENERIC:
+                    return new GenericLocalizationProvider();
+                case I18N_WARNING:
+                    return new WarningLocalizationProvider(resolutionOrder);
+                default:
+                    return this.languageLocalizationProvidersByLanguage.get(providerName);
+            }
+        }));
+    }
+
+
+}
+
+export class I18nNamespace {
+    public readonly namespacePath: ReadonlyArray<string>;
+    private readonly typeI18n: ReadonlyArray<TypeI18n>;
+    private readonly fieldI18n: ReadonlyArray<FieldI18n>;
+    constructor(input: FlatNamespaceI18nConfig) {
+        this.namespacePath = input.namespacePath;
+        this.typeI18n = this.extractTypes(input);
+        this.fieldI18n = this.extractFields(input);
+    }
+
+    public getAllLocalizationsForType(name: string) {
+        return this.typeI18n.find(typeTrans => typeTrans.name === name);
+    }
+
+    public getAllLocalizationsForField(name: string, type: string|undefined): ReadonlyArray<FieldI18n> {
+        return compact([
+            this.fieldI18n.find(fieldTrans => fieldTrans.name === name && fieldTrans.type === type),
+            this.fieldI18n.find(fieldTrans => fieldTrans.name === name && fieldTrans.type === undefined)
+        ]);
+    }
+
+    private extractFields(input: FlatNamespaceI18nConfig): ReadonlyArray<TypeI18n> {
+        return [
+            // Namespace fields
+            ...Object.keys(input.fields).map(fieldName =>
+                ({
+                    name: fieldName,
+                    label: input.fields[fieldName].label,
+                    hint: input.fields[fieldName].hint
+                })
+            ),
+            // Fields from types
+            ...flatMap(Object.keys(input.types), typeName =>
+                Object.keys(input.types[typeName].fields)
+                    .map(fieldName =>
+                        ({
+                            name: fieldName,
+                            label: input.types[typeName].fields[fieldName].label,
+                            hint: input.types[typeName].fields[fieldName].hint,
+                            type: typeName
+                        })
+                    )
+            )
+        ];
+    }
+
+    private extractTypes(input: FlatNamespaceI18nConfig): ReadonlyArray<FieldI18n> {
+        return Object.keys(input.types).map(typeName => ({
+                name: typeName,
+                singular: input.types[typeName].singular,
+                plural: input.types[typeName].plural,
+                hint: input.types[typeName].hint
+            })
+        );
+    }
+
+}
+
+export interface TypeLocalization {
+    readonly singular?: string,
+    readonly plural?: string,
+    readonly hint?: string
+}
+
+export interface FieldLocalization {
+    readonly label?: string,
+    readonly hint?: string,
+}
+
+export interface TypeI18n extends TypeLocalization {
```

Comment:
I think the terms I18n and Localization are used confusingly here. What is this name? Maybe this interface should be internal?